packages/backend/src/open_payments/auth/middleware.ts

```diff
@@ -55,26 +53,19 @@ export function createAuthMiddleware({
           ctx.throw(401, `Invalid signature`)
         }
       }
-      await GrantReference.transaction(async (trx: Transaction) => {
-        const grantRef = await grantReferenceService.get(grant.grant, trx)
-        if (grantRef) {
-          if (grantRef.clientId !== grant.clientId) {
-            logger.debug(
-              `clientID ${grant.clientId} for grant ${grant.grant} does not match internal reference clientId ${grantRef.clientId}.`
-            )
-            ctx.throw(500)
-          }
-        } else if (action === AccessAction.Create) {
-          // Grant and client ID's are only stored for create routes
-          await grantReferenceService.create(
-            {
-              id: grant.grant,
-              clientId: grant.clientId
-            },
-            trx
-          )
-        }
-      })
+
+      try {
+        await grantReferenceService.getOrCreate(
+          { id: grant.grant, clientId: grant.clientId },
+          action
+        )
```

Comment:
```suggestion
        if (action === AccessAction.Create) {
          await grantReferenceService.getOrCreate({
            id: grant.grant,
            clientId: grant.clientId
          })
        } else {
          await grantReferenceService.get(grant.grant)
        }
```

Reply:
By moving the action check to the caller, the `clientId` check 
https://github.com/interledger/rafiki/blob/c6eaaa91825bc840d0b6bd4aa10b9b3e4cb1896c/packages/backend/src/open_payments/grantReference/service.ts#L58-L63
will probably need to be moved to the `getGrantRerefence` method
https://github.com/interledger/rafiki/blob/c6eaaa91825bc840d0b6bd4aa10b9b3e4cb1896c/packages/backend/src/open_payments/grantReference/service.ts#L27-L29
and will have to take both arguments (`id` and `clientId`)
```ts
async function getGrantReference(
  options: CreateGrantReferenceOptions,
  trx: Transaction
) {
// ...
}
```
otherwise there will be no validation for the `clientId` when calling `grantReferenceService.get`

Reply:
I'm leaning towards not passing `clientId` to `grantReferenceService.get` (even though this is currently the only place calling it).
What if we moved the `clientId` check out here?
```suggestion
        if (action === AccessAction.Create) {
          await grantReferenceService.getOrCreate({
            id: grant.grant,
            clientId: grant.clientId
          })
        } else {
          const grantRef = await grantReferenceService.get(grant.grant)
          if (grantRef && grantRef.clientId !== grant.clientId) {
            throw new Error(
              `clientID ${grant.clientId} for grant ${grant.id} does not match internal reference clientId ${grantRef.clientId}.`
            }
          }
        }
```
:thinking: that looks a lot like how it was before...

Is there an argument for not actually using `getOrCreate`?
It looks like the `AccessAction.Create` check was added after this issue was opened:
- https://github.com/interledger/rafiki/pull/676